src/app/components/common-ui/WebForm/index.js

```diff
@@ -0,0 +1,335 @@
+import _ from 'underscore';
+import cn from 'classnames';
+import op from 'object-path';
+import PropTypes from 'prop-types';
+
+import React, {
+    forwardRef,
+    useEffect,
+    useLayoutEffect as useWindowEffect,
+    useImperativeHandle,
+    useRef,
+    useState,
+} from 'react';
+
+const useLayoutEffect =
+    typeof window !== 'undefined' ? useWindowEffect : useEffect;
+
+const noop = () => {};
+
+const ENUMS = {
+    STATUS: {
+        COMPLETE: 'COMPLETE',
+        READY: 'READY',
+        SENDING: 'SENDING',
+        ERROR: 'ERROR',
+    },
+};
+
+/**
+ * -----------------------------------------------------------------------------
+ * Hook Component: WebForm
+ * -----------------------------------------------------------------------------
+ */
+let WebForm = (props, ref) => {
+    const {
+        className,
+        namespace,
+        required,
+        onBeforeSubmit,
+        onComplete,
+        onError,
+        onSubmit,
+        onUpdate,
+        showError,
+        validator,
+        value,
+        name,
+        id,
+        children,
+        ...formProps
+    } = props;
+
+    // Refs
+    const formRef = useRef();
+    const stateRef = useRef({
+        value,
+        error: null,
+        mounted: false,
+        status: ENUMS.STATUS.READY,
+        elements: {},
+    });
+
+    // State
+    const [, setNewState] = useState(stateRef.current);
+
+    // Internal Interface
+    const setState = newState => {
+        // Update the stateRef
+        stateRef.current = {
+            ...stateRef.current,
+            ...newState,
+        };
+
+        // Trigger useEffect()
+        setNewState(stateRef.current);
+    };
+
+    // External Interface
+    useImperativeHandle(ref, () => ({
+        setState,
+        state: stateRef.current,
+        update,
+    }));
+
+    // Side Effects
+    useLayoutEffect(() => {
+        update(value);
+    }, Object.values(value));
+
```

Comment:
Shorten? 
😂_me being a well actually guy because I think maybe you miss it from Updox_

```
useLayoutEffect(() => update(value), Object.values(value));
```

Reply:
I like it shorter that way. 😆